### DISCUSSION TITLE
Tuist란?

### DISCUSSION THREAD

**Discussion by JUNY0110** [Samsamhada/MacC-Team-Samsamhada#197] [General]
## Tuist
Xcode 프로젝트와의 상호 작용, 유지 관리, 생성을 용이하게하는 것을 목표로하는 CLI
Binary로 배포되어 다른 tool에 의존하지 않고도 쉽게 설치 및 사용 가능.

## Tuist 설치
```terminal
curl -Ls https://install.tuist.io | bash
```
터미널에 그대로 입력.
<details>
<summary>  <h3>xcrun error발생 시</h3> </summary>
<div markdown="1">

예시 이미지)
<img width="574" alt="스크린샷 2022-12-23 오후 10 02 18" src="https://user-images.githubusercontent.com/98405970/209340578-f65c9d73-df91-4805-bc2f-f51b6c0aa843.png">

```terminal
xcode-select --install
```
터미널에 그대로 입력.
- 이유: MacOS 업데이트 후, git을 포함한 다수의 개발 툴과 관련해 **Command Line Tools를 식별하지 못하는 문제가 발생**하기 때문에 해당 명령어를 통해 Command Line developer Tools를 설치해주면 해당 문제가 해결이 됨.
<img width="498" alt="스크린샷 2022-12-23 오후 10 09 58" src="https://user-images.githubusercontent.com/98405970/209341613-1ca75a9e-3d0f-4462-85ae-82b51867549e.png">

이후 정상적으로 설치되는 것을 확인할 수 있음.
<img width="572" alt="스크린샷 2022-12-23 오후 10 11 03" src="https://user-images.githubusercontent.com/98405970/209341737-78b7a2a9-921d-4406-8180-bb110a901199.png">

</div>
</details>

<details>
<summary>  <h3>Tuist 설정 및 프로젝트 생성</h3> </summary>
<div markdown="1">

Tuist 프로젝트 생성
```terminal
mkdir MyApp
cd MyApp
```
```terminal
tuist init --platform ios                             // UIKit
tuist init --platform ios --template swiftui          // SwiftUI
```

Tuist 설정 변경
```terminal
tuist edit
--> 반영 후 ctrl + C를 눌러 종료
```

생성된 Project.swift파일을 아래와 같이 수정 ( 프로젝트 이름: `MyApp`은 tuist 설치 간 작성한 임시 이름임 )
```swift
import ProjectDescription
import ProjectDescriptionHelpers
import MyPlugin

let project = Project(
    name: "MyApp", 
    targets: [
        Target(
            name: "MyApp",
            platform: .iOS,
            product: .app,
            bundleId: "io.tuist.MyApp",
            infoPlist: .default,
            sources: ["Targets/MyApp/Sources/**"]
        ),
        Target(
            name: "MyAppTests",
            platform: .iOS,
            product: .unitTests,
            bundleId: "io.tuist.MyAppTests",
            infoPlist: .default
        )
    ]
)
```

Tuist 반영된 프로젝트 생성
```terminal
tuist generate
```

</div>
</details>

**Comment by ikanic** [comment]
Project의 구성요소

- name: String 타입, 앱의 이름, ex. `MyApp`
- organizationName: String? 타입, organization 이름, 삼삼하다의 경우 비워두면 됨
- options: Project.Options 타입
- packages: [Package] 타입
- settings: Settings? 타입
- targets: [Target] 타입
- schemes: [Scheme] 타입
- fileHeaderTemplate: FileHeaderTemplate? 타입
- additionalFiles: [FileElement] 타입
- resourceSynthesizers: [ResourceSynthesizer] 타입

**Comment by ikanic** [comment]
Target의 구성요소

- name: String 타입, 타겟 이름, ex. `MyApp`, `MyAppTest`
- platform: Platform 타입, 타겟 플랫폼, ex. `.iOS`, `.macOS`, `.tvOS`, `.watchOS`
- product: Product 타입
- productName: String? 타입
- bundleId: String 타입, 번들 ID, ex. `com.immeenu.myapp`
- deploymentTarget: DeploymentTarget? 타입, ex. `.iOS(targetVersion: "16.2", devices: [.iphone, .ipad])`
- infoPlist: InfoPlist? 타입, info.plist
- sources: SourceFilesList? 타입
- resources: ResourceFileElements? 타입
- copyFiles: [CopyFilesAction]? 타입
- headers: Headers? 타입
- entitlements: Path? 타입
- scripts: [TargetScript] 타입
- dependencies: [TargetDependency] 타입
- settings: Settings? 타입
- coreDataModels: [CoreDataModel] 타입
- environment: [String: String] 타입
- launchArguments: [LaunchArgument] 타입
- additionalFiles: [FileElement] 타입
